### ISSUE TITLE
application logs missing failed actions

### ISSUE THREAD

**Issue by tduehr** [digitalocean/netbox#1834]
### Issue type
[ ] Feature request 
[X] Bug report      
[ ] Documentation   

### Environment
* Python version:  3.5.2
* NetBox version:  2.2.8

### Description
The console/file log output of Netbox is limited to authentication events.

I'm currently trying to debug the API as used by NetboxClientRuby. I'm getting 403 errors where I don't think I should be when creating objects (the API token has full write privileges and my user can create objects in the web UI fine). Looking at the logs in `/var/log/supervisor`, all I see are my authentication events. Further, turning the logging up to `DEBUG` does not change the output.

`403 Forbidden ({"detail"=>"You do not have permission to perform this action."})`

I expect `DEBUG` level logging to include errors in object creation for any reason. Especially, 403s which should likely show up at `INFO` if not all the way up at `ERROR`.

**Comment by jeremystretch** [comment]
Authentication errors are just that: errors. There's nothing else to log because as soon as authentication fails, the request is aborted. What are you asking be changed?

**Comment by tduehr** [comment]
In general, netbox lacks robust logging. At an INFO level I expect all authentication attempts and all successful actions to be logged to the console/syslog/etc.. I need to be able to audit netbox as it is (or was) used at a given time. The only way to reliably do that is through logs.

**Comment by pm17788** [comment]
I'd like to add another vote to this request.
Right now, I'm wiring a bunch of Python stuff to populate data in NetBox, but since I'm doing this via the API, there does not seem to be a good audit log for it. 

I would argue that only the modify actions should be auditable, though, at DEBUG levels, where the implication of the performance impact is at least implied, even read APIs should be logged.

**Comment by jeremystretch** [comment]
Still not clear on what you're asking for. Could you please provide example code that would accomplish what you're proposing?

**Comment by pm17788** [comment]
@jeremystretch: Apologies. I am not a decent enough coder to actually do that. 

I'll try to rephrase for more sanity. We are talking about taking, effectively, what is the "Recent Activity" feed and putting it out to a log (be it a file or SYSLOG). The change not currently there is also logging the API access(es).

Specifically, instead of:
`2018-01-30 14:09 | Modified device server515`

The log line would contain:

If modified via WebUI:
`{"datetime":"2018-01-30T14:12:03-05:00", "user":"pm17788", "interface":"WebUI", "operation":"modify", "id":"dcim/devices/515", "data":{"name":"server515"}}`
If modified via API:
`{"datetime":"2018-01-30T14:12:03-05:00", "user":"pm17788", "interface":"API", "operation":"modify", "id":"dcim/devices/515", "data":{"name":"server515"}}`

Basically, a full audit log of not only the operation performed, but what were the actual data items associated with the operation. If, for example, an Interface was created, then a full JSON of the data passed is what I'd expect in the logline.

It (probably) doesn't make sense to store this sort of an audit trail in a database - long-term storage requirements get silly, but IFF feasible, having more detail in the Recent Activity log would make sense for, say, past 7 days. If I want longer retention, I get to setup separate logging.

I hope this makes sense.

**Comment by jeremystretch** [comment]
The UserAction stuff is slated to be overhauled entirely at some point (hopefully soon). #1127 has been open for a while now to convert it to use signals, but we need to go further than that. I might just expand that issue to cover this.

**Comment by pm17788** [comment]
Cool! I would be quite willing to test, if that helps in any way.

**Comment by jeremystretch** [comment]
Folding this into #1898
